pgosupport.cmake

```diff
@@ -19,25 +28,57 @@ function(add_pgo TargetName)
     )
 
     # Enable PGO only for optimized configs
-    set(ConfigTypeList RELEASE RELWITHDEBINFO)
+    set(ConfigList RELEASE RELWITHDEBINFO)
+    set(IsReleaseConfig "$<OR:$<CONFIG:RELEASE>,$<CONFIG:RELWITHDEBINFO>>")
 
-    foreach(ConfigType IN LISTS ConfigTypeList)
-        set(LinkFlagsProperty "LINK_FLAGS_${ConfigType}")
-        if(CLR_CMAKE_PGO_INSTRUMENT)
+    if(CLR_CMAKE_PGO_INSTRUMENT)
+        # Unfortunately LINK_FLAGS_* don't support generator expressions, so we need to use a loop
+        foreach(Config IN LISTS ConfigList)
             if(WIN32)
-                set_property(TARGET ${TargetName} APPEND_STRING PROPERTY ${LinkFlagsProperty} "/LTCG /GENPROFILE")
+                append_prop_strings(${TargetName} LINK_FLAGS_${Config} /LTCG /GENPROFILE)
+            elseif(UNIX)
+                append_prop_strings(${TargetName} LINK_FLAGS_${Config} -flto -fuse-ld=gold -fprofile-instr-generate)
             endif(WIN32)
-        else(CLR_CMAKE_PGO_INSTRUMENT)
-            # If we don't have profile data availble, gracefully fall back to a non-PGO opt build
-            if(EXISTS ${ProfilePath})
+        endforeach(Config)
+        if(UNIX)
+            # On Unix we need to pass PGO flags to the compiler as well as the linker
+            target_compile_options(${TargetName} PRIVATE $<${IsReleaseConfig}:-flto -fprofile-instr-generate>)
+        endif(UNIX)
+    else(CLR_CMAKE_PGO_INSTRUMENT)
+        # If we don't have profile data availble, gracefully fall back to a non-PGO opt build
+        # Likewise, if PGO is not supported, gracefully fall back to a non-PGO opt build
+        if(CLR_CMAKE_HAVE_PGO AND EXISTS ${ProfilePath})
+            # Unfortunately LINK_FLAGS_* don't support generator expressions, so we need to use a loop
+            foreach(Config IN LISTS ConfigList)
                 if(WIN32)
-                    set_property(TARGET ${TargetName} APPEND_STRING PROPERTY ${LinkFlagsProperty} "/LTCG /USEPROFILE:PGD=${ProfilePath}")
+                    append_prop_strings(${TargetName} LINK_FLAGS_${Config} /LTCG /USEPROFILE:PGD=${ProfilePath})
+                elseif(UNIX)
+                    append_prop_strings(${TargetName} LINK_FLAGS_${Config} -flto -fuse-ld=gold -fprofile-instr-use=${ProfilePath})
                 endif(WIN32)
-            endif(EXISTS ${ProfilePath})
-        endif(CLR_CMAKE_PGO_INSTRUMENT)
-    endforeach(ConfigType)
+            endforeach(Config)
+            if(UNIX)
+                ## On Unix we need to pass PGO flags to the compiler as well as the linker
+                target_compile_options(${TargetName} PRIVATE $<${IsReleaseConfig}:-flto -fprofile-instr-use=${ProfilePath}>)
```

Comment:
Since the makefile generator is not a multi-target one, you could just use `if (UPPERCASE_CMAKE_BUILD_TYPE STREQUAL RELEASE OR UPPERCASE_CMAKE_BUILD_TYPE STREQUAL RELWITHDEBINFO)`, which we do in other places. Unless something leaked in without my knowledge, we use the generator expressions for WIN32 only, which uses a multi-target generator.